src/java/org/jivesoftware/openfire/roster/Roster.java

```diff
@@ -625,7 +625,7 @@ public void broadcastPresence(Presence packet) {
             }
         }
         // Broadcast presence to shared contacts whose subscription status is FROM
-        for (String contact : implicitFrom.keySet()) {
+        for (String contact : ((Map<String,Set<String>>) implicitFrom).keySet()) { // Cast prevents Java compatibility issue as described in OF-1116. Remove the cast when Java 7 support is dropped from Openfire.

```

Comment:
I'd prefer to change the type of the local variable `implicitFrom` to `Map`. It's cleaner imo (see http://stackoverflow.com/q/147468) and should fix the problem as well, according to https://gist.github.com/AlainODea/1375759b8720a3f9f094

We also don't need to revert anything.